pytorch_lightning/profiler/profilers.py

```diff
@@ -282,3 +285,217 @@ def __del__(self):
+            return
+
+        self.profiler.__exit__(
+            exc_type=None,
+            exc_val=None,
+            exc_tb=None
+        )
+
+        events = self.profiler.function_events
+        self.profiler = None
+        for name in self.running_stack:
+            if name not in self.profiled_actions:
+                self.profiled_actions[name] = events
+            else:
+                self.profiled_actions[name] += events
+
+    def stop(self, action_name: str) -> None:
+        if action_name in self.profiled_functions:
+            if len(self.running_stack) == 0 or self.running_stack[-1] != action_name:
+                raise ValueError(  # pragma: no-cover
+                    f"Attempting to stop recording an action ({action_name}) which was never started."
+                )
+            self._stop(action_name)
+            self.running_stack.pop()
+            # restore running profiler
+            if len(self.running_stack) > 0:
+                self._start(self.running_stack[-1])
+
+    def summary(self) -> str:
+        recorded_stats = {}
+        output_string = ''
+
+        if self.enabled:
+            for action_name, events in self.profiled_actions.items():
+
+                # next line is a workaround for a pytorch issue (fixed on master, still present
+                # on 1.7). Without it the code fails with `AssertionError: There is already a CPU
+                # parent event for detach`
+                events.populate_cpu_children = lambda: None
+
+                if self.export_to_chrome:
+                    filename = f"{action_name}_trace.json"
+                    path_to_trace = filename if self.path_to_export_trace is None \
+                        else os.path.join(self.path_to_export_trace, filename)
+                    events.export_chrome_trace(path_to_trace)
+
+                if self.emit_nvtx:
+                    return output_string
+
+                else:
+                    table = events.key_averages(
+                        group_by_input_shapes=self.group_by_input_shapes).table(
+                            sort_by=self.sort_by_key,
+                            row_limit=self.row_limit)
+                    recorded_stats[action_name] = table
+
+            # log to standard out
+            output_string = f"{os.linesep}Profiler Report{os.linesep}"
+            for action, stats in recorded_stats.items():
+                output_string += (
+                    f"{os.linesep}Profile stats for: {action}{os.linesep}{stats}"
+                )
+
+        return output_string
+
+    def describe(self):
+        """Logs a profile report after the conclusion of the training run."""
+        super().describe()
+        if self.output_file:
+            self.output_file.flush()
+
+    def __del__(self):
+        """Close profiler's stream."""
+        if self.output_file:
+            self.output_file.close()
+
+    @property
+    def available_sort_by_keys(self):
```

Comment:
1) why is this a property? Just seems like a weird design, shouldn't it be in the docst instead?
2) similar thing for `profiled_functions` would be nice

Reply:
Agree. Removed !